From 4c3c6afb82fb26c51900f041686cfb7157147d6f Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Wed, 20 Aug 2025 15:55:45 +0800
Subject: [PATCH 232/236] mtk: hostapd: add new function ap_cleanup_sta()

The purpose of the ap_cleanup_sta() function is to safely clean up the
data structures of a specific STA at the MLD level.

This function first clears the WLAN_STA_ASSOC flag from the sta_info on
all links. As a result, ap_get_sta() will always return the sta_info
belonging to the input hapd, or NULL. After that, it proceeds to remove
all sta_info entries on all links.

This function is especially useful when, due to certain errors, the
sta_info of the setup link is accidentally deleted while the sta_info of
other links remain. Using this function allows for safely cleaning up
the entire STA MLD.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/sta_info.c | 40 ++++++++++++++++++++++++++++++++++++++++
 src/ap/sta_info.h |  1 +
 2 files changed, 41 insertions(+)

diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index 014be4912..6bf11cec4 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -2043,3 +2043,43 @@ bool ap_sta_in_list(struct hostapd_data *hapd, struct sta_info *sta) {
 	return false;
 }
 
+void ap_cleanup_sta(struct hostapd_data *hapd, u8 *addr) {
+	struct hostapd_data *h;
+	struct sta_info *sta;
+
+	sta = ap_get_sta(hapd, addr);
+	if (sta)
+		sta->flags &= ~WLAN_STA_ASSOC;
+
+#ifdef CONFIG_IEEE80211BE
+	if (hostapd_is_mld_ap(hapd)) {
+		/* Clear the flag WLAN_STA_ASSOC so that ap_get_sta() must
+		 * return the sta_info of the input hapd or NULL.
+		 */
+		for_each_mld_link(h, hapd) {
+			if (h == hapd)
+				continue;
+
+			sta = ap_get_sta(h, addr);
+			if (!sta)
+				continue;
+
+			sta->flags &= ~WLAN_STA_ASSOC;
+		}
+
+		for_each_mld_link(h, hapd) {
+			if (h == hapd)
+				continue;
+
+			sta = ap_get_sta(h, addr);
+			if (!sta)
+				continue;
+
+			ap_free_sta(h, sta);
+		}
+	}
+#endif /* CONFIG_IEEE80211BE */
+	sta = ap_get_sta(hapd, addr);
+	if (sta)
+		ap_free_sta(hapd, sta);
+}
diff --git a/src/ap/sta_info.h b/src/ap/sta_info.h
index 92674ecb0..707f1e7ab 100644
--- a/src/ap/sta_info.h
+++ b/src/ap/sta_info.h
@@ -437,5 +437,6 @@ void hostapd_free_link_stas(struct hostapd_data *hapd);
 void clear_wpa_sm_for_each_partner_link(struct hostapd_data *hapd,
 					struct sta_info *psta);
 bool ap_sta_in_list(struct hostapd_data *hapd, struct sta_info *sta);
+void ap_cleanup_sta(struct hostapd_data *hapd, u8 *addr);
 
 #endif /* STA_INFO_H */
-- 
2.45.2

